<a href="https://colab.research.google.com/github/1289nav/model-inversion-database-reconstruction/blob/main/Database_Reconstruction_Attack.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install adversarial-robustness-toolbox

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 14.7 MB/s eta 0:00:00


In [ ]:
pip install scikit-learn==1.1.0 diffprivlib

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.7/30.7 MB 33.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 176.5/176.5 kB 22.4 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.2.2
    Uninstalling scikit-learn-1.2.2:
      Successfully uninstalled scikit-learn-1.2.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
bigframes 1.0.0 requires scikit-learn>=1.2.2, but you have scikit-learn 1.1.0 which is incompatible.


In [ ]:
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

from art.attacks.inference.reconstruction import DatabaseReconstruction

import numpy as np

import pandas as pd

import matplotlib.pyplot as plt


# Task 1: Logistic regression on Iris dataset

## Training Model

In [ ]:
dataset = datasets.load_iris()

# Convert the dataset to a Pandas DataFrame
df = pd.DataFrame(data=np.c_[dataset['data'], dataset['target']], columns=dataset['feature_names'] + ['target'])

#Displaying the dataset
df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
..      ...  
145     2.0  
146     2.0  
147     2.0  
148     2.0  
149     2.0  

[150 rows x 5 columns]

In [ ]:
X=dataset['data']
y=dataset['target']

#Scaling the data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=37)

In [ ]:
# Initialize model
iris_reg = LogisticRegression(max_iter=5000)

In [ ]:
# Training the Logistic regression model
iris_reg.fit(X_train, y_train)

LogisticRegression(max_iter=5000)

In [ ]:
# Predict the labels for the test set
y_pred = iris_reg.predict(X_test)

# Calculate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy of Logistic Regression on Iris dataset:", accuracy)

Accuracy of Logistic Regression on Iris dataset: 1.0


## Reconstruction

In [ ]:
from art.estimators.classification.scikitlearn import ScikitlearnLogisticRegression

#Creating an ART estimator for logistic regression
art_iris_reg = ScikitlearnLogisticRegression(iris_reg)

In [ ]:
#Defining a database reconstruction instance based on Sci Kit learn class
dbrecon = DatabaseReconstruction(art_iris_reg)

In [ ]:
num_sample_recon1=0
recon1_state=[]
recon1_RMSE=[]

for i in range(X_train.shape[0]):
  print("i= ",i)
  victim=i
  X_train_adversary=np.delete(X_train,victim,axis=0)
  y_train_adversary=np.delete(y_train,victim,axis=0)

  x, y = dbrecon.reconstruct(X_train_adversary, y_train_adversary)

  if(np.argmax(y) == y_train[victim]):
    num_sample_recon1+=1
    print("Success")
    recon1_state.append(1)
  else:
    print("Failed")
    recon1_state.append(0)

  RMSE=np.sqrt(((X_train[victim] - x) ** 2).sum() / X_train.shape[1])
  recon1_RMSE.append(RMSE)
  print("Inference RMSE: ",RMSE)

i=  0
i=  1
i=  2
i=  3
i=  4
i=  5
i=  6


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr

i=  7
i=  8
i=  9
i=  10
i=  11
i=  12
i=  13
i=  14
i=  15
i=  16
i=  17
i=  18
i=  19
i=  20
i=  21
i=  22
i=  23
i=  24
i=  25
i=  26
i=  27
i=  28
i=  29
i=  30
i=  31
i=  32
i=  33
i=  34
i=  35
i=  36
i=  37
i=  38
i=  39
i=  40
i=  41
i=  42
i=  43
i=  44
i=  45
i=  46
i=  47
i=  48
i=  49
i=  50
i=  51
i=  52
i=  53
i=  54
i=  55
i=  56
i=  57
i=  58
i=  59
i=  60
i=  61
i=  62
i=  63
i=  64
i=  65
i=  66
i=  67
i=  68
i=  69
i=  70
i=  71
i=  72
i=  73
i=  74
i=  75
i=  76
i=  77


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr

i=  78
i=  79
i=  80
i=  81
i=  82
i=  83
i=  84
i=  85
i=  86
i=  87
i=  88
i=  89
i=  90
i=  91
i=  92
i=  93
i=  94
i=  95
i=  96
i=  97
i=  98
i=  99
i=  100
i=  101
i=  102
i=  103
i=  104
i=  105
i=  106
i=  107
i=  108
i=  109
i=  110
i=  111
i=  112
i=  113
i=  114
i=  115
i=  116
i=  117
i=  118
i=  119
i=  120


IndexError: index 120 is out of bounds for axis 0 with size 120

In [ ]:
print("Total number of samples tested for reconstruction: ", 120)
print("Total number reconstructed success  : ",num_sample_recon)
print("Success rate: ", (num_sample_recon/120)*100, " %")

Total number of samples tested for reconstruction:  120
Total number reconstructed success  :  87
Success rate:  72.5  %


# Logistic Regression on a 2 class classification dataset

## Training the model

In [ ]:
#For the 2 class classification problem we will be using the
# Breast cancer Wisconcin (Diagnostic) dataset
# which is a binary classification problem

from sklearn.datasets import load_breast_cancer

#Load breast cancer dataset
cancer_dset=load_breast_cancer()

# Create a DataFrame from the dataset's data and target attributes
df2=pd.DataFrame(data=cancer_dset.data, columns=cancer_dset.feature_names)

# Add target labels to the DataFrame
df2['target'] = cancer_dset.target

df2

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

In [ ]:
print(df2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [ ]:
X2=cancer_dset['data']
y2=cancer_dset['target']

#Scaling the data
scaler = StandardScaler()
X2_scaled = scaler.fit_transform(X2)

# Train test split
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y2, test_size=0.2, random_state=37)

In [ ]:
# Initialize model
cancer_reg = LogisticRegression(max_iter=5000)

In [ ]:
# Training the Logistic regression model
cancer_reg.fit(X2_train, y2_train)

LogisticRegression(max_iter=5000)

In [ ]:
# Predict the labels for the test set
y2_pred = cancer_reg.predict(X2_test)

# Calculate the accuracy of the model
accuracy2 = accuracy_score(y2_test, y2_pred)
print("Accuracy of Logistic Regression on cancer dataset:", accuracy2)

Accuracy of Logistic Regression on cancer dataset: 0.9736842105263158


## Reconstruction attack

In [ ]:
#We first generate a random array with no repeated entries to randomly select a victim data point in the dataset
#Due to computational constraints we will run this for 50 different victims

import random
# Generate a random array of 50 entries without repetition in the range of 1 to 564
size=X2_train.shape[0]
print(size)
victim_array = random.sample(range(0, size), 10)

print(victim_array)

455
[265, 14, 173, 401, 134, 28, 327, 64, 80, 215]


In [ ]:
#Creating an ART estimator for logistic regression
art_cancer_reg = ScikitlearnLogisticRegression(cancer_reg)

#Defining a database reconstruction instance based on Sci Kit learn class
dbrecon2 = DatabaseReconstruction(art_cancer_reg)

In [ ]:
num_sample_recon_2=0
recon2_state=[]
recon_RMSE=[]

for a in victim_array :
  print("victim Id= ",a)
  victim=a
  X_train_adversary=np.delete(X2_train,victim,axis=0)
  y_train_adversary=np.delete(y2_train,victim,axis=0)

  x, y = dbrecon2.reconstruct(X_train_adversary, y_train_adversary)

  if(np.argmax(y) == y2_train[victim]):
    num_sample_recon_2+=1
    recon2_state.append(1)
    print("Success")
  else:
    recon2_state.append(0)
    print("Failed")

  RMSE=np.sqrt(((X2_train[victim] - x) ** 2).sum() / X2_train.shape[1])
  recon_RMSE.append(RMSE)
  print("Inference RMSE: ",RMSE)

victim Id=  265
Success
Inference RMSE:  0.4831757786107144
victim Id=  14
Success
Inference RMSE:  1.8936971675598848
victim Id=  173
Success
Inference RMSE:  0.49820215677948027
victim Id=  401
Success
Inference RMSE:  1.2928776133810398
victim Id=  134
Success
Inference RMSE:  0.9417396971422878
victim Id=  28
Failed
Inference RMSE:  5.8593455482656065
victim Id=  327
Failed
Inference RMSE:  2.3823201364960496
victim Id=  64
Success
Inference RMSE:  0.18706597763379423
victim Id=  80
Success
Inference RMSE:  0.8838421862450201
victim Id=  215
Failed
Inference RMSE:  1.8738037955050377


In [ ]:
print("Total number of samples tested for reconstruction: ", 10)
print("Total number reconstructed success  : ",num_sample_recon_2)
print("Success rate: ", (num_sample_recon_2/10)*100, " %")
print("The average reconstruction accuraccy is: ", sum(recon_RMSE)/len(recon_RMSE))

Total number of samples tested for reconstruction:  10
Total number reconstructed success  :  7
Success rate:  70.0  %
The average reconstruction accuraccy is:  1.6296070057618917


In most situations the reconstructions RMSE is around 1

Though not very high, it could be lower indicating some level of uncertainity still remains around the predicted data points after reconstruction attack. The possible reasons could be
1. The model is complex with several features. The dataset contains 30 different features, making accurate reconstruction more difficult
2. However the prediction success rate is good. This indicates that the reconstructed samples still fall in the ranges of the respective classes. This indicates an acceptable fit of the model
3. In most situations the reconstructed data points and prediction on the same matches the original victim. However in the cases it doesnt, its possible our reconstructed data point is not accurate.
Eg: This particularly noticed for victim 28 and 327 which have reconstructions RMSE greater than 2 which is considerably high

However overall the reconstruction attack has been sucessful with a good success rate

# Four Class classification dataset

In [ ]:
from sklearn.datasets import make_classification

rand_x, rand_y = make_classification(n_samples=1000,n_informative=6,n_redundant=4,n_classes=4,random_state=42)

In [ ]:
#Displaying the generated synthetic dataset
df3=pd.DataFrame(data=rand_x)
df3['target']=rand_y

df3

0         1          2         3         4         5         6  \
0   -0.322549 -1.046034  -1.889299  0.125559  0.046919  1.228728  0.841082   
1    1.202249  0.443940  -2.319904 -0.413449 -0.825012  0.269053  1.206536   
2    0.861098  0.607865  -2.966707  1.881621 -0.200658  1.569584 -0.623180   
3   -1.980836 -0.464936  -4.044210  0.414784  2.714319  0.818766 -0.038324   
4    0.024432  1.114819  -3.975394 -1.951603  1.065195  1.002236  0.320998   
..        ...       ...        ...       ...       ...       ...       ...   
995  0.358741  2.147944  -3.321257 -1.129486  1.904014  1.149477  1.157422   
996  0.283053  2.713434  -1.405938 -0.802430 -0.829173 -0.041337 -0.339311   
997  0.103538 -0.974978 -11.557259  0.269181  3.377539 -0.260036  1.431152   
998  0.995303 -0.215136  -2.047148  0.407879  1.312312 -0.305632  0.814731   
999  0.366606  0.214213  -4.101404  0.808360 -1.228810  1.746491 -0.795718   

            7         8         9  ...        11        12        13  \
0   -1.044187 -0.740295  1.206846  ...  1.865727 -0.552219 -1.184014   
1    1.213179 -1.732728  2.264312  ... -0.617286 -0.030637  0.760730   
2   -0.668433 -0.730327  1.206742  ...  1.018625 -1.112410 -0.073518   
3   -0.480574  0.760321 -1.949747  ... -1.157416 -3.672651 -2.354950   
4    1.912005 -0.808056  2.274254  ...  1.349179 -0.726160  0.336149   
..        ...       ...       ...  ...       ...       ...       ...   
995  0.171039  1.599867 -1.988617  ... -1.696316 -3.045777  0.190564   
996 -0.368619 -0.373053 -0.079102  ...  1.119688 -0.799656  1.741952   
997 -0.269952 -5.162097  6.218132  ...  1.594178 -2.764034 -1.329020   
998 -0.961472 -0.461427  1.953557  ...  0.687764  0.224179 -0.142151   
999 -1.259065 -2.002223  3.152150  ... -0.723168 -0.735376  0.001408   

           14        15        16        17        18        19  target  
0    0.144765  2.342857  0.661313 -0.433775  0.847929  0.296681       2  
1   -0.924268 -1.451226 -1.509681 -2.035419  0.532892  3.842231       2  
2   -0.853218  0.051076 -0.046387 -1.959953  4.479084  2.568818       3  
3   -0.678980 -2.396082  2.399123  0.353581 -1.874892 -3.422538       1  
4   -0.051542 -1.066032  0.071889 -2.784730  1.164060  2.633473       0  
..        ...       ...       ...       ...       ...       ...     ...  
995  1.547311 -1.348428 -0.622085 -2.236573  0.133679  0.270643       3  
996 -0.074927 -0.348157  0.235299 -1.646437  0.807046  3.498691       0  
997  0.245039 -0.287906  0.312873 -3.666996 -1.543451  3.492119       0  
998 -0.212860 -1.360024 -0.488317 -1.458965 -0.051713  0.531077       1  
999 -1.067434 -0.002238 -0.952034 -3.132375 -1.005315  5.112246       2  

[1000 rows x 21 columns]

In [ ]:
# Train test split
X3_train, X3_test, y3_train, y3_test = train_test_split(rand_x, rand_y, test_size=0.2, random_state=37)

In [ ]:
# Generate 100 victim data points
size=X3_train.shape[0]
print(size)
victim_array_3 = random.sample(range(0, size), 100)

print(victim_array_3)

800
[528, 352, 578, 600, 512, 585, 248, 122, 495, 169, 695, 383, 350, 369, 402, 798, 477, 172, 282, 582, 100, 55, 761, 63, 658, 549, 429, 754, 473, 740, 30, 236, 53, 423, 526, 340, 386, 226, 401, 17, 301, 84, 759, 275, 601, 554, 156, 756, 729, 496, 735, 517, 131, 261, 244, 427, 68, 725, 132, 92, 406, 324, 483, 652, 766, 253, 206, 193, 262, 632, 698, 421, 3, 294, 13, 109, 263, 280, 521, 734, 319, 538, 69, 220, 375, 598, 584, 181, 514, 228, 101, 739, 618, 216, 594, 544, 405, 480, 224, 705]


##Case 1: Logistic regression


In [ ]:
# Initialize model
rand_reg = LogisticRegression(max_iter=10000)

In [ ]:
# Training the Logistic regression model
rand_reg.fit(X3_train, y3_train)

LogisticRegression(max_iter=10000)

In [ ]:
# Predict the labels for the test set
y3A_pred = rand_reg.predict(X3_test)

# Calculate the accuracy of the model
accuracy3 = accuracy_score(y3_test, y3A_pred)
print("Accuracy of Logistic Regression on noisy dataset:", accuracy3)

Accuracy of Logistic Regression on noisy dataset: 0.68


In [ ]:
#Creating an ART estimator for logistic regression
art_rand_reg = ScikitlearnLogisticRegression(rand_reg)

#Defining a database reconstruction instance based on Sci Kit learn class
dbrecon3_reg = DatabaseReconstruction(art_rand_reg)

In [ ]:
num_sample_recon_3_reg=0
recon3_reg_state=[]
recon3_reg_RMSE=[]

for a in victim_array_3 :
  print("victim Id= ",a)
  victim=a
  X_train_adversary=np.delete(X3_train,victim,axis=0)
  y_train_adversary=np.delete(y3_train,victim,axis=0)

  x, y = dbrecon3_reg.reconstruct(X_train_adversary, y_train_adversary)

  if(np.argmax(y) == y3_train[victim]):
    num_sample_recon_3_reg+=1
    recon3_reg_state.append(1)
    print("Success")
  else:
    recon3_reg_state.append(0)
    print("Failed")

  RMSE=np.sqrt(((X3_train[victim] - x) ** 2).sum() / X3_train.shape[1])
  recon3_reg_RMSE.append(RMSE)
  print("Inference RMSE: ",RMSE)

victim Id=  528
Success
Inference RMSE:  0.5272721452782172
victim Id=  352
Success
Inference RMSE:  3.118929931613295
victim Id=  578
Success
Inference RMSE:  1.4842987199834718
victim Id=  600
Success
Inference RMSE:  0.28957466030956336
victim Id=  512


KeyboardInterrupt: 


## Case 2: Naive Baye's

In [ ]:
import sklearn.naive_bayes as naive_bayes

In [ ]:
rand_NB = naive_bayes.GaussianNB().fit(X3_train, y3_train)

In [ ]:
print("Model accuracy (on the test dataset): {}".format(rand_NB.score(X3_test, y3_test)))

Model accuracy (on the test dataset): 0.625


# New Section